### ISSUE TITLE
How to disconnect specified client from outside hub

### ISSUE THREAD

**Issue by bluetianx** [aspnet/SignalR#2866]
I have a web Api for disconnectting specified client ,how to disconnect specified client from outside hub.
I do not find a method about disconnect client from IHubContext<T>;

**Comment by atresnjo** [comment]
Yeah IClientProxy doesn't have the Abort method. It's just available in the hub context. I'd suggest invoking a method on client side which will gracefully disconnect then.

**Comment by anurse** [comment]
We don't really support this from HubContext, but it's something to consider for the future. A graceful shutdown based on invoking a "Shutdown" method on the client is more appropriate here, as @atresnjo suggests. Abruptly terminating connections should generally only be used when the client is unresponsive for uncooperative.

It's difficult to add this ability to the HubContext because in a multi-server scenario (when scaling), we don't know if the code that's calling "Abort" is running on the same server as the connection. If it's not, then we have to send that information to the appropriate server, which has a whole bunch of other issues behind it.

**Comment by bluetianx** [comment]
I want client to   invok  OnDisconnectedAsync method (`public override async Task OnDisconnectedAsync (Exception exception)` ) of Hub class. Is this a best practice ?

**Comment by atresnjo** [comment]
@bluetianx 

No. Implement your own method and invoke it when needed.

Something like:

    public class ServerHub : Hub
    {
    }

    public class ClientHub : Hub
    {
        public ClientHub()
        {
            hubConnection.On<string>("Shutdown",
                ShutdownReceived);
        }

        public void ShutdownReceived(string reason)
        {

        }
    }

    public class ShutdownService
    {
        private readonly IHubContext<ServerHub> _hubContext;
        public ShutdownService(IHubContext<ServerHub> hub)
        {
            _hubContext = hub;
        }

        public void SendShutdown()
        {
            _hubContext.Clients.All.SendAsync("Shutdown", "Reason");
        }
    }

**Comment by BrennanConroy** [comment]
Do not call `OnDisconnectedAsync` manually. It will not do what you expect.
`OnDisconnectedAsync` is called when the connection is terminated. So in your `Shutdown` method you can call `Context.Abort();` which will then trigger the `OnDisconnectedAsync` method correctly.

**Comment by anurse** [comment]
>     connection.on("Shutdown", function (message) {
>         connection.invoke("Shutdown", message);
>     });
>     
>     connection.on("Disconnect", function (message) {
>         connection.invoke("OnDisconnectedAsync", null);
>     });

In these cases, you can just use `connection.stop` on the client to disconnect. There's no reason to call `OnDisconnectedAsync`. And, as @BrennanConroy said, it will not work to call it manually.

Here's a complete sample that will allow you to signal a client to disconnect from the server **outside the Hub**:

NotificationHub.cs:
```csharp
public class NotificationHub: Hub
{
    // You don't need any code here to specifically support this. But if you want to
    // run code when a client does disconnect, you can override OnDisconnectedAsync
}
```

Client.js
```javascript
var connection = new signalR.HubConnectionBuilder().withUrl("http://localhost:5050/notificationhub").build();

connection.on("RequestShutdown", function (reason) {
    console.log("Server requested we disconnect because: " + reason);
    connection.stop();
});
    
connection.start().catch(function (err) {
    console.log(err.toString());
    return console.error(err.toString());
}).then(function(data) {
    // Do startup things
});
```

WebAPI.cs
```csharp
[Route("api/[controller]")]
[ApiController]
public class ValuesController : ControllerBase
{
    private readonly IHubContext<NotificationHub> _hubContext;

    [HttpPost]
    public bool DisconnectUser(string userName)
    {
        _hubContext.User(userName).SendAsync("RequestShutdown", "reason");
    }
}
```

**Comment by bluetianx** [comment]
OK,I understand,Thank you very much @atresnjo @anurse @BrennanConroy

**Comment by anurse** [comment]
Blocked by #3284
